File: wind-composer-web/src/music/bassEngine.ts
import type { MelodyNoteDto } from "../types";
import { clamp } from "../utils";
import type { MusicMemory } from "./musicMemory";
import type { StyleProfile } from "./styleEngine";

export interface BassContext {
  chordTones: number[];
  energy: number;
  windDirection: number;
  pressureTrend: number;
  barInPhrase: number;
  style: StyleProfile;
}

export class BassEngine {
  constructor(private memory: MusicMemory) {}

  onBar(ctx: BassContext): MelodyNoteDto[] {
    if (!ctx.chordTones.length) return [];
    const root = ctx.chordTones[0];
    const fifth = ctx.chordTones[Math.min(2, ctx.chordTones.length - 1)];
    const style = ctx.style.bassStyle;

    let pattern: number[];
    if (style === "ambient") {
      pattern = [root - 12];
    } else if (style === "house") {
      pattern = this.housePattern(root, fifth, ctx);
    } else if (style === "trance") {
      pattern = [root - 12, root - 5, root - 12, fifth - 12];
    } else if (style === "techno") {
      pattern = [root - 12, root - 12, fifth - 12, root - 12];
    } else {
      pattern = [root - 12, fifth - 12];
    }

    if (this.memory.bassOverused(pattern)) {
      pattern = pattern.length > 1 ? [...pattern].reverse() : pattern;
    }
    this.memory.rememberBass(pattern);

    const vel = clamp(0.38 + ctx.energy * 0.48);
    return pattern.map((midi, i) => ({
      midi,
      velocity: vel * (0.88 + i * 0.03),
      duration_sec: 0.32 + (style === "house" ? 0.05 : 0),
    }));
  }

  private housePattern(root: number, fifth: number, ctx: BassContext): number[] {
    if (ctx.barInPhrase % 4 === 0) return [root - 12, root - 12, fifth - 12, root - 12];
    if (Math.random() < 0.35 + ctx.energy * 0.22) {
      return [root - 12, root - 7, fifth - 12, root - 12];
    }
    return [root - 12, root - 12, root - 12, fifth - 12];
  }
}

File: wind-composer/bass_engine.py
"""Procedural bass line generation."""

from __future__ import annotations

import random
from dataclasses import dataclass
from typing import List, Optional

from melody_engine import MelodyNote
from music_memory import MusicMemory
from style_engine import StyleProfile
from utils import clamp


@dataclass
class BassContext:
    chord_tones: List[int]
    energy: float
    wind_direction: float
    pressure_trend: float
    bar_in_phrase: int
    style: StyleProfile


class BassEngine:
  def __init__(self, memory: MusicMemory) -> None:
    self._memory = memory
    self._pattern_idx = 0
    self._last_root: Optional[int] = None

  def on_bar(self, ctx: BassContext) -> List[MelodyNote]:
    if not ctx.chord_tones:
      return []
    root = ctx.chord_tones[0]
    fifth = ctx.chord_tones[min(2, len(ctx.chord_tones) - 1)]
    style = ctx.style.bass_style

    if style == "ambient":
      pattern = (root - 12,)
    elif style == "house":
      pattern = self._house_pattern(root, fifth, ctx)
    elif style == "trance":
      pattern = (root - 12, root - 5, root - 12, fifth - 12)
    elif style == "techno":
      pattern = (root - 12, root - 12, fifth - 12, root - 12)
    else:
      pattern = (root - 12, fifth - 12)

    if self._memory.bass_overused(pattern):
      pattern = tuple(reversed(pattern)) if len(pattern) > 1 else pattern

    self._memory.remember_bass(pattern)
    vel = clamp(0.35 + ctx.energy * 0.45)
    notes: List[MelodyNote] = []
    for i, midi in enumerate(pattern):
      notes.append(MelodyNote(midi=midi, velocity=vel * (0.9 + i * 0.02), duration_sec=0.35))
    return notes

  def _house_pattern(self, root: int, fifth: int, ctx: BassContext) -> tuple:
    if ctx.bar_in_phrase % 4 == 0:
      return (root - 12, root - 12, fifth - 12, root - 12)
    if random.random() < 0.35 + ctx.energy * 0.2:
      return (root - 12, root - 7, fifth - 12, root - 12)
    return (root - 12, root - 12, root - 12, fifth - 12)
